### ISSUE TITLE
Chroot doesn't work with busybox

### ISSUE THREAD

**Issue by errordeveloper** [ansible/ansible#16097]
With busybox, `/bin/sh` can be a relative symlink, and won't resolve correctly prior to `chroot()`.

```
PLAY ***************************************************************************

TASK [setup] *******************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: OSError: [Errno 2] No such file or directory: '/rootfs/bin/sh'
fatal: [/rootfs]: FAILED! => {"failed": true, "stdout": ""}

NO MORE HOSTS LEFT *************************************************************

PLAY RECAP *********************************************************************
/rootfs                    : ok=0    changed=0    unreachable=0    failed=1   
```

```
# ls -la /rootfs/bin/sh
lrwxrwxrwx 1 root root 12 Jun  1 22:17 /rootfs/bin/sh -> /bin/busybox
```

A sensible solution for this would be through an option to override [the hardcoded shell path](https://github.com/ansible/ansible/blob/v2.0.0.1-1/lib/ansible/plugins/connection/chroot.py#L67-L68), or something like that.

**Comment by bcoca** [comment]
`ansible_executable` host var or `executable` in ansible.cfg already do that, they just need to be added to the chroot plugin

**Comment by abadger** [comment]
I don't think ansible_executable can be used here since that's used from inside of the target and this is being used from outside of the target.

**Comment by abadger** [comment]
The code looks like it is just sanity checking that the chroot is going to be usable by us.  Perhaps we should change the check to simply pass if chroot + "/bin/sh" is a symlink instead of an executable.  ansible_executable could probably be added into the mix as well but it would be solving a separate problem than is being reported here.

**Comment by errordeveloper** [comment]
This sounds right to me.

**Comment by errordeveloper** [comment]
Thank you, Toshio.

**Comment by abadger** [comment]
@errordeveloper give that Pull Request a try and see if it solves the problem for you.

**Comment by jktr** [comment]
This doen't seem completely fixed with #16239, as I ran into this issue today.
```
# ansible --version
ansible 2.3.2.0
  config file = /root/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jul 21 2017, 03:24:34) [GCC 7.1.1 20170630]
```

I'm unsure if this is the right place to post this, but a bug in the above fix seems to be as follows:

75beca2 from #16239 uses ansible's `is_executable` as the first check, which throws `OSError` if the chroot's `/bin/sh` is something like
```
# ll /var/lib/lxc/irc/rootfs/bin/ |grep -w sh
lrwxrwxrwx 1 root root   12 17-09-17 20:22 sh -> /bin/busybox
```
This is due to `is_executable` trying to derefere the symlink outside the chroot's context and, unless the host provides `/bin/busybox`, failing with:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 125, in run
    res = self._execute()
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 472, in _execute
    self._connection = self._get_connection(variables=variables, templar=templar)
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 719, in _get_connection
    connection = self._shared_loader_obj.connection_loader.get(conn_type, self._play_context, self._new_stdin)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/__init__.py", line 378, in get
    obj = obj(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/chroot.py", line 71, in __init__
    if not (is_executable(chrootsh) or (os.path.lexists(chrootsh) and os.path.islink(chrootsh))):
  File "/usr/lib/python2.7/site-packages/ansible/module_utils/basic.py", line 570, in is_executable
    return ((stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH) & os.stat(path)[stat.ST_MODE])
OSError: [Errno 2] No such file or directory: '/var/lib/lxc/irc/rootfs/bin/sh'
```
It seems wrong to dereference the symlink outside the chroot context here.

A possible solution might be pulling the symlinks check in front of `is_executable` and short-circuiting.
This gives something like the following, which seems to work for me:
```
if not ((os.path.lexists(chrootsh) and os.path.islink(chrootsh)) or is_executable(chrootsh)):
    raise AnsibleError("%s does not look like a chrootable dir (/bin/sh missing)" % self.chroot)
```
